### ISSUE TITLE
Switch to TensorFlow 2.0 and tf.keras

### ISSUE THREAD

**Issue by de-code** [kermitt2/delft#87]
From the [Keras site](https://keras.io/):

> The current release is Keras 2.3.0, which makes significant API changes and add support for TensorFlow 2.0. The 2.3.0 release will be the last major release of multi-backend Keras. Multi-backend Keras is superseded by `tf.keras`.

Maybe it's time to upgrade?

**Comment by lfoppiano** [comment]
I'd say yes! but after #82

**Comment by kermitt2** [comment]
Hello ! I've already started to look at it (it's in the TODO list at the end of the README), but it's complicated due to TF dependencies to BERT and ELMo (also the FLAIR implementation I am exploring at https://github.com/kensho-technologies/bubs).

**Comment by de-code** [comment]
Any more up-to-date thoughts on this? I am not sure about the current or near future support situation of TensorFlow 1.x. It seems on its way out. For some things there are backwards compatible APIs.

**Comment by kermitt2** [comment]
Still no ELMo compatible TF2, so not really possible. For BERT there's now https://github.com/kpe/bert-for-tf2 but it would probably involve quite a lot of effort for a change, because it's a rewrite from scratch and I had to introduced a lot of modifications linked to the original BERT to get the sequence labeling working efficiently and with a CRF layer (the original BERT repo was not supporting sequence labeling).  

Last release of TF 1.15 is 3 months ago, so I am not worry for the moment :)

**Comment by de-code** [comment]
> Last release of TF 1.15 is 3 months ago, so I am not worry for the moment :)

From the [TensorFlow 1.15.0 release](https://github.com/tensorflow/tensorflow/releases/tag/v1.15.0) (Oct 16, 2019):

> This is the last 1.x release for TensorFlow. We do not expect to update the 1.x branch with features, although we will issue patch releases to fix vulnerabilities for at least one year.

That is now more than a year. I didn't find the planned support for it.

On one of our repos we got the automatic low severity security notifications:

- [CVE-2020-15266](https://github.com/advisories/GHSA-xwhf-g6j5-j5gc)
- [CVE-2020-15265](https://github.com/advisories/GHSA-rrfp-j2mp-hq9c)

Those were fixed in [TensorFlow 2.4.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.4.0) (Dec 14, 2020). There wasn't a related TensorFlow 1.x release with those fixes. As far as I can tell, the security advisories do not mention it not affecting 1.x (just that it affects `< 2.4.0`).

Additionally TensorFlow 1.15.4 doesn't support Python 3.8+ ([according to PyPI](https://pypi.org/project/tensorflow/1.15.4/)). Whereas TensorFlow 2.4.0 supports 3.8, but not Python 2 ([see PyPI](https://pypi.org/project/tensorflow/2.4.0/)).

Which makes me think that it may not be far off before the support is officially dropped.

**Comment by kermitt2** [comment]
Ok now I am slowly starting to worry :D 

I think there's no real solution except skipping ELMo entirely or moving to pytorch, which would be almost like starting a new project.

**Comment by de-code** [comment]
> I think there's no real solution except skipping ELMo entirely or moving to pytorch, which would be almost like starting a new project.

I am not using ELMo (or BERT) at the moment. I guess you are using it? What do we use for ELMo, I didn't see it in the requirements. In general the upgrade shouldn't be too difficult in particular when using the compatibility layer first. Converting it to be proper TensorFlow 2.0 might require more of a re-think.

Yes, switching to pytorch should probably be a separate project. I am not sure what the benefits would be for this project.

**Comment by de-code** [comment]
BTW pandas `1.2.0` dropped support for Python 3.6 ([release notes](https://pandas.pydata.org/pandas-docs/version/1.2.0/whatsnew/v1.2.0.html), [GitHub issue](https://github.com/pandas-dev/pandas/pull/35214)). Narrowing the Python version to 3.7, supported by both pandas 1.2.0+ and TensorFlow 1.x (if you did decide to keep your other dependencies up-to-date).

**Comment by de-code** [comment]
Any thought of adding TensorFlow 2.x support any time soon?

**Comment by kermitt2** [comment]
I am hoping to work on this and on ONNX support during the end of year holidays.

**Comment by de-code** [comment]
Closing the issue so silently? It almost deserves a big announcement. Well done!
